### PR TITLE
no CryptKeys unless necessary in localize, and put KeyFinder in the context CORE-4187

### PR DESCRIPTION
**Pull request by mmaxim**
This patch does the following:

1.) Remove the call to `GetTLFCryptKeys` in `localizeConversation` if the TLF name in question does not have an @ sign in it (chance to be an SBS name). The canonical name is inserted in `PostLocal`, so there is no chance of some random name being in there.
2.) Remove the `chat/utils` package and merge it into the `chat` package.
3.) Remove `UserInfoMapper`
4.) Add a new function `chat.Context` which creates a custom chat context that contains the identify mode, identify break result array, and a `KeyFinder` object. This allows us to more aggressively use the `KeyFinder` as we wind through the various code paths and reduce calls to `GetTLFCryptKeys`. It also reduces the amount of code that tried to call into `GetTLFCryptKeys` manually.

cc @patrickxb

### CODE REVIEW

File: go/service/tlf.go [hunk 1]
```diff
@@ -100,14 +106,18 @@ func (h *tlfHandler) CryptKeys(ctx context.Context, arg keybase1.TLFQuery) (keyb
 }
 
 func (h *tlfHandler) PublicCanonicalTLFNameAndID(ctx context.Context, arg keybase1.TLFQuery) (keybase1.CanonicalTLFNameAndIDWithBreaks, error) {
+	var err error
+	ident, breaks, ok := chat.IdentifyMode(ctx)
+	if ok {
+		arg.IdentifyBehavior = ident
+	}
+	defer h.G().Trace(fmt.Sprintf("tlfHandler.PublicCanonicalTLFNameAndID(tlf=%s,mode=%v)", arg.TlfName,
```
Comment (maxtaco) [inline go/service/tlf.go]: Use `CTrace` and thread through the `context.Context`?

File: go/service/tlf.go [hunk 2]
```diff
@@ -78,15 +79,20 @@ func (h *tlfHandler) sendNotifyEvent(update keybase1.CanonicalTLFNameAndIDWithBr
...
 func (h *tlfHandler) CryptKeys(ctx context.Context, arg keybase1.TLFQuery) (keybase1.GetTLFCryptKeysRes, error) {
+	var err error
+	ident, breaks, ok := chat.IdentifyMode(ctx)
+	if ok {
+		arg.IdentifyBehavior = ident
+	}
+	defer h.G().Trace(fmt.Sprintf("tlfHandler.CryptKeys(tlf=%s,mode=%v)", arg.TlfName,
+		arg.IdentifyBehavior), func() error { return err })()
 
+	debug.PrintStack()
```
Comment (maxtaco) [inline go/service/tlf.go]: wrap this in a `call-site` VDEBUG-style logger?

File: go/service/tlf.go [hunk 3]
```diff
@@ -78,15 +79,20 @@ func (h *tlfHandler) sendNotifyEvent(update keybase1.CanonicalTLFNameAndIDWithBr
 }
 
 func (h *tlfHandler) CryptKeys(ctx context.Context, arg keybase1.TLFQuery) (keybase1.GetTLFCryptKeysRes, error) {
+	var err error
+	ident, breaks, ok := chat.IdentifyMode(ctx)
+	if ok {
+		arg.IdentifyBehavior = ident
+	}
+	defer h.G().Trace(fmt.Sprintf("tlfHandler.CryptKeys(tlf=%s,mode=%v)", arg.TlfName,
```
Comment (maxtaco) [inline go/service/tlf.go]: again use `CTrace`?

File: go/chat/inboxsource.go [hunk 4]
```diff
@@ -414,15 +409,18 @@ func (s *localizer) localizeConversation(ctx context.Context, uid gregor1.UID,
 		return chat1.ConversationLocal{Error: &errMsg}
 	}
 
-	info, err := utils.LookupTLF(ctx, s.getTlfInterface(), conversationLocal.Info.TlfName, conversationLocal.Info.Visibility)
-	if err != nil {
-		errMsg := err.Error()
-		return chat1.ConversationLocal{Error: &errMsg}
+	// Only do this check if there is a chance the TLF name might be an SBS name.
+	if strings.Contains(conversationLocal.Info.TlfName, "@") {
```
Comment (mmaxim) [inline go/chat/inboxsource.go]: Here is the change for not calling `CryptKeys` during `localize`

Comment (patrickxb) [inline go/chat/inboxsource.go]: colon is also an option:

    cd /keybase/private/patrick,twitter:malgorithms

works.

Comment (patrickxb) [inline go/chat/inboxsource.go]: follow up:  there's a lot of code in libkb/assertion.go that we should probably use instead of strings.Contains("@").

**Pull request merged**